### PR TITLE
Add hot reload button and support service registration.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Uses the existing 'reloadSources' service in flutter_tools. Will implement hot restart in a later CL.

Piggy backs off of @jacob314 's demo.

### CODE REVIEW

File: lib/service_manager.dart [hunk 1]
```diff
@@ -51,6 +52,30 @@ class ServiceConnectionManager {
 
   Stream<Null> get onConnectionClosed => _connectionClosedController.stream;
 
+  Future<Response> callService(String name, {Map args}) {
```
Comment (jacob314) [inline lib/service_manager.dart]: Add a comment describing the restriction that there must be only one registered provider of the service.

File: lib/service_manager.dart [hunk 2]
```diff
@@ -62,6 +87,14 @@ class ServiceConnectionManager {
       }
 
       this.service = service;
+
+      service.onServiceEvent.listen((e) {
+        if (e.kind == EventKind.kServiceRegistered) {
+          methodsForService.putIfAbsent(e.service, () => []).add(e.method);
+          print(e);
```
Comment (jacob314) [inline lib/service_manager.dart]: remove print statement

Comment (kenzieschmoll) [reply lib/service_manager.dart]: done.

File: lib/service_manager.dart (hunk 1 above)
Comment (kenzieschmoll) [reply lib/service_manager.dart]: done.

File: lib/ui/ui_utils.dart [hunk 3]
```diff
@@ -69,3 +69,14 @@ CoreElement createExtensionCheckBox(String extensionName) {
         text,
       ]));
 }
+
+// TODO(kenzieschmoll): add hotRestart button, register method in flutter_tools.
+
+CoreElement createHotReloadButton() {
+  final PButton button = new PButton('Hot Reload')..small();
+  button.click(() async {
+    await serviceManager.callMulticastService('reloadSources',
```
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply lib/ui/ui_utils.dart]: Done.

[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply lib/ui/ui_utils.dart]: done.

File: lib/service_manager.dart [hunk 4]
```diff
@@ -51,6 +52,30 @@ class ServiceConnectionManager {
...
+          '${registered.length}');
+    }
+    return service.callMethod(registered.first, args: args);
+  }
+
+  /// Call a service that may have been registered by multiple clients.
+  ///
+  /// For example, a service to navigate a code editor to a specific line and
+  /// column might be registered by multiple code editors.
+  Future<List<Response>> callMulticastService(String name, {String isolateId, Map args}) {
```
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply lib/service_manager.dart]: Done.

File: lib/service_manager.dart [hunk 5]
```diff
@@ -51,6 +52,33 @@ class ServiceConnectionManager {
...
+  /// column might be registered by multiple code editors.
+  Future<List<Response>> callMulticastService(String name,
+      {String isolateId, Map args}) {
+    final registered = methodsForService[name] ?? const [];
+    if (registered.isNotEmpty) {
+      return Future.wait(registered.map((String method) {
+        return service.callMethod(method, isolateId: isolateId, args: args);
+      }));
+    } else {
+      throw Exception('There are no registered methods for service $name');
```
Comment (jacob314) [inline lib/service_manager.dart]: nit: maybe put $name in quotes or put a colon after the word service.

File: test/service_manager_test.dart [hunk 6]
```diff
@@ -108,6 +108,101 @@ void main() {
...
+          Exception('Expected one registered service for fakeMethod'
+              ' but found 0');
+
+      try {
+        await serviceManager.callService('fakeMethod');
+      } catch (e) {
+        exception = e;
+      }
+
+      expect(exception, isNotNull);
```
Comment (jacob314) [inline test/service_manager_test.dart]: package:test can make this logic a little cleaner to write.
For example you should be able to write
expect(service.manager.callService('fakeMethod'), throwsA(typeTypeItThrows));
https://pub.dartlang.org/documentation/test/latest/index.html#asynchronous-tests
    expect(new Future.error("oh no"), throwsA(equals("oh no")));

File: test/service_manager_test.dart [hunk 7]
```diff
@@ -108,6 +108,101 @@ void main() {
...
+
+    test('callMulticastService', () async {
+      dynamic _e;
+
+      final registeredService =
+          serviceManager.methodsForService['reloadSources'] ?? const [];
+      expect(registeredService, isNotEmpty);
+
+      try {
+        await serviceManager.callMulticastService('reloadSources',
```
Comment (jacob314) [inline test/service_manager_test.dart]: nit: you can verify that a valid Result object is returned as well.

File: test/service_manager_test.dart [hunk 8]
```diff
@@ -108,6 +108,101 @@ void main() {
...
+    test('hotReload', () async {
+      dynamic _e;
+
+      try {
+        await serviceManager.performHotReload();
+      } catch (e) {
+        _e = e;
+      }
+
+      expect(_e, isNull);
```
Comment (jacob314) [inline test/service_manager_test.dart]: you can simplify these tests by just writing
result = await serviceManager.performHotReload();
expect(result.json[foo], equals(bar));
if the method throws an exception the test will fail so no need for the catch block.

File: lib/service_manager.dart [hunk 9]
```diff
@@ -102,6 +138,17 @@ class ServiceConnectionManager {
     _stateController.add(null);
     _connectionClosedController.add(null);
   }
+
+  // TODO(kenzie): add hot restart method, register method in flutter_tools.
```
Comment (devoncarew) [inline lib/service_manager.dart]: @aam would be a good one to coordinate with here

File: lib/service_manager.dart [hunk 10]
```diff
@@ -102,6 +138,17 @@ class ServiceConnectionManager {
...
+
+  Future<void> performHotReload() async {
+    try {
+      await callMulticastService('reloadSources',
+          isolateId: _isolateManager.selectedIsolate.id);
+    } catch (e) {
+      print('Error during hot reload: "$e."');
```
Comment (devoncarew) [inline lib/service_manager.dart]: Let's add a TODO here to improve error handling (we need a better way for the app in general, and then we can call it from places in the code like this one).

File: lib/ui/ui_utils.dart [hunk 11]
```diff
@@ -69,3 +69,15 @@ CoreElement createExtensionCheckBox(String extensionName) {
         text,
       ]));
 }
+
+// TODO(kenzie): add hotRestart button.
+
+CoreElement createHotReloadButton() {
```
Comment (devoncarew) [inline lib/ui/ui_utils.dart]: This is a very specific method - unlikely to be used outside of a specific library. Perhaps move this to the library where it'll be used?

Comment (devoncarew) [reply lib/ui/ui_utils.dart]: That library may not yet exist? In which case lets just add a todo to relocate this method later.

File: lib/service_manager.dart (hunk 5 above)
Comment (kenzieschmoll) [reply lib/service_manager.dart]: added quotes.

File: lib/service_manager.dart (hunk 10 above)
Comment (kenzieschmoll) [reply lib/service_manager.dart]: done.

File: lib/ui/ui_utils.dart (hunk 11 above)
Comment (kenzieschmoll) [reply lib/ui/ui_utils.dart]: Added a todo. Likely will use this button in the timeline lib.

File: lib/service_manager.dart [hunk 12]
```diff
@@ -51,6 +52,33 @@ class ServiceConnectionManager {
 
   Stream<Null> get onConnectionClosed => _connectionClosedController.stream;
 
+  /// Call a service that is registered by exactly one client.
+  Future<Response> callService(String name, {String isolateId, Map args}) {
```
Comment (jacob314) [inline lib/service_manager.dart]: make this method async

Comment (kenzieschmoll) [reply lib/service_manager.dart]: done.

File: test/service_manager_test.dart (hunk 6 above)
Comment (kenzieschmoll) [reply test/service_manager_test.dart]: done

File: test/service_manager_test.dart (hunk 7 above)
Comment (kenzieschmoll) [reply test/service_manager_test.dart]: done.

File: test/service_manager_test.dart (hunk 8 above)
Comment (kenzieschmoll) [reply test/service_manager_test.dart]: done.

File: lib/service_manager.dart (hunk 9 above)
Comment (jacob314) [reply lib/service_manager.dart]: Btw @aam helpe me figure out how to call this service extension from the devtools web app.

File: test/service_manager_test.dart [hunk 13]
```diff
@@ -109,98 +109,39 @@ void main() {
...
-      try {
-        await serviceManager.callService('reloadSources',
-            isolateId: serviceManager.isolateManager.selectedIsolate.id);
-      } catch (e) {
-        _e = e;
-      }
-
-      expect(_e, isNull);
+      await serviceManager.callService('reloadSources',
+          isolateId: serviceManager.isolateManager.selectedIsolate.id);
```
Comment (jacob314) [inline test/service_manager_test.dart]: nit: add a trailing comma to get the flutter style for these calls which I think looks a little nicer.

Comment (kenzieschmoll) [reply test/service_manager_test.dart]: done.

**Pull request merged**